lib/vdsm/storage/fileVolume.py

```diff
@@ -552,11 +554,13 @@ def _truncate_volume(cls, vol_path, capacity, vol_id, dom):
             raise
 
     @classmethod
-    def _allocate_volume(cls, vol_path, size, preallocate):
+    def _allocate_volume(
+            cls, vol_path, size, preallocate, vol_format, cow_compat=None):
         try:
             # Always create sparse image, since qemu-img create uses
             # posix_fallocate() which is inefficient and harmful.
-            op = qemuimg.create(vol_path, size=size, format=qemuimg.FORMAT.RAW)
+            op = qemuimg.create(
+                vol_path, size=size, format=vol_format, qcow2Compat=cow_compat)
```

Comment:
This creates a sparse image with virtual_size=size...

```
$ qemu-img create -f qcow2 /data/tmp/test.qcow2 1g
Formatting '/data/tmp/test.qcow2', fmt=qcow2 cluster_size=65536 extended_l2=off compression_type=zlib size=1073741824 lazy_refcounts=off refcount_bits=16
```

```
$ ls -lhs /data/tmp/test.qcow2
196K -rw-r--r--. 1 nsoffer nsoffer 193K Aug 27 17:43 /data/tmp/test.qcow2
```

But in line 577 we run:

```
            # If the image is preallocated, allocate the rest of the image
            # using fallocate helper. qemu-img create always writes zeroes to
            # the first block so we should skip it during preallocation.
            if size > 4096 and preallocate == sc.PREALLOCATED_VOL:
                op = fallocate.allocate(vol_path, size - 4096, offset=4096)
```

Which write zeros to the image after offset 4096, overwriting data in offset 4k-196k, corrupting the qcow2 image.

Lets try this:

```
$ helpers/fallocate $((1024*1024*1024 - 4096)) --offset 4096 /data/tmp/test.qcow2 
```

```
$ ls -lhs /data/tmp/test.qcow2
1.1G -rw-r--r--. 1 nsoffer nsoffer 1.0G Aug 27 17:47 /data/tmp/test.qcow2
```

```
$ qemu-img info /data/tmp/test.qcow2 
image: /data/tmp/test.qcow2
file format: qcow2
virtual size: 1 GiB (1073741824 bytes)
disk size: 1 GiB
cluster_size: 65536
Format specific information:
    compat: 1.1
    compression type: zlib
    lazy refcounts: false
    refcount bits: 16
    corrupt: false
    extended l2: false
```

```
$ qemu-img check /data/tmp/test.qcow2 
No errors were found on the image.
Image end offset: 262144
```

So we were lucky, and there was no data after offset 4096, but this is unsafe. Future version of qcow2 can write data after offset 4096, and this change will corrupt the image.

We can make this safe using:

    offset = qemuimg.check(vol_path, format="qcow2")["offset"]

    # qemu-img offset is usually not aligned to 4k, and fallocate uses
    # direct I/O requiring alignment to logical block size. Using 4k
    # alignment ensures this works for any storage.
    offset = utils.round(offset, 4 * KiB)

    op = fallocate.allocate(vol_path, size - offset, offset=offset)
    ...

Reporting image end offset in "qemu-img check" is public API that is safe to use for truncating or reducing block devices to the end of the image, so it will always be safe for allocating additional space after the end of the image.

This works only for qcow2 format, for raw format st_size is always the virtual size,
and we cannot overwrite any data. Actually we could always allocated the entire size since new raw empty image must be empty.

But fixing this issue by modifying the code to create raw volume is the wrong approach. You may introduce regressions in creating raw images by adding support for preallocating qcow2, and the code becomes complicated and hard to understand. The best way is to have separate functions (allocate_raw_volume, allocatte_qcow2_volume) with simple and clear logic, and share some low level helpers like (run_fallocate_task) which both functions need.

Reply:
Ah, is true. I should be more careful with the qcow2 format, and its header :facepalm: 

I will try what you proposed and post it.

Alternatively, we may want to consider using the preallocation `falloc` in the `qemu-img create` call: https://github.com/oVirt/vdsm/blob/master/lib/vdsm/storage/qemuimg.py#L218

This comment states that is inefficient: https://github.com/oVirt/vdsm/blob/master/lib/vdsm/storage/fileVolume.py#L557
But I found [benchmarks](https://www.jamescoyle.net/how-to/2060-qcow2-physical-size-with-different-preallocation-settings) where only the `full` setting seems to be really slow. `falloc` is similar to what we do, but (I assume) already taking into account the format and qcow2 header.

Reply:
preallocation=falloc is fine on modern storage but harmful on NFS 3, this is why we don't use it and we have our own fallocate helper. Check the commits adding our fallocate helper to understand why.